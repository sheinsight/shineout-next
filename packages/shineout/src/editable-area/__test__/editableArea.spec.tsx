import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import EditableArea from '..';
import mountTest from '../../tests/mountTest';
import { classLengthTest } from '../../tests/structureTest';
import {
  attributesTest,
  baseTest,
  classContentTest,
  createClassName,
  displayTest,
  snapshotTest,
  styleContentTest,
  styleTest,
  textContentTest,
} from '../../tests/utils';
import EditableAreaBase from '../__example__/01-base';
import EditableAreaControlled from '../__example__/02-controlled';
import EditableAreaContainer from '../__example__/03-container';
import EditableAreaRenderFooter from '../__example__/t-04-renderFooter';
import EditableAreaResult from '../__example__/t-05-renderResult';

const SO_PREFIX = 'editableArea';
const originClasses = ['wrapper', 'place', 'clear', 'placeholder', 'content', 'popup'];
const originItemClasses = [
  'wrapperDisabled',
  'wrapperNoBorder',
  'wrapperError',
  'wrapperSmall',
  'wrapperLarge',
  'wrapperPaddingBox',
  'wrapperInnerTitle',
  'wrapperInnerTitleTop',
  'wrapperInnerTitleBottom',
  'popupShow',
];
const {
  wrapper,
  place,
  clear,
  content,
  popup,
  placeholder,
  wrapperDisabled,
  wrapperNoBorder,
  wrapperPaddingBox,
  wrapperInnerTitleTop,
  wrapperInnerTitleBottom,
  popupShow,
} = createClassName(SO_PREFIX, originClasses, originItemClasses);

const textareaFooterClassName = '.so-textarea-footer';

const changeValueHandle = (container: HTMLElement, value: string) => {
  fireEvent.click(container.querySelector(place)!);
  fireEvent.change(document.querySelectorAll('textarea')[0], { target: { value } });
  fireEvent.blur(document.querySelectorAll('textarea')[0]);
};

const defaultValue: string = 'defaultValue';
const value: string = 'value';

const testValue = (container: HTMLElement, contentValue: string) => {
  textContentTest(container.querySelector(content)!, contentValue);
  textContentTest(document.querySelectorAll('textarea')[0], contentValue);
  textContentTest(document.querySelectorAll('textarea')[1], contentValue);
};

afterEach(cleanup);
mountTest(<EditableArea />);

describe('EditableArea[Base]', () => {
  displayTest(EditableArea, 'ShineoutEditableArea');
  baseTest(EditableArea, wrapper);
  snapshotTest(<EditableAreaBase />);
  snapshotTest(<EditableAreaControlled />);
  snapshotTest(<EditableAreaContainer />);
  snapshotTest(<EditableAreaRenderFooter />);
  snapshotTest(<EditableAreaResult />);
  test('should render default', () => {
    const { container } = render(<EditableArea />);
    const editableWrapper = container.querySelector(wrapper)!;
    classContentTest(editableWrapper, wrapperNoBorder);
    const editablePlace = editableWrapper.querySelector(place)!;
    classContentTest(editablePlace, wrapperPaddingBox);
    classContentTest(editablePlace, wrapperInnerTitleTop);
    classContentTest(editablePlace, wrapperInnerTitleBottom);
    attributesTest(editablePlace, 'tabindex', '0');
    const editableContent = editablePlace.querySelector(content)!;
    expect(editableContent?.querySelector(placeholder)).toBeInTheDocument();
    const editablePopup = document.querySelector(popup)!;
    expect(editablePopup).toBeInTheDocument();
    classLengthTest(editablePopup, 'textarea', 2);
    fireEvent.click(editablePlace);
    classContentTest(editablePopup, popupShow);
  });
  test('should render when set bordered is true', () => {
    const { container } = render(<EditableArea bordered />);
    const editableWrapper = container.querySelector(wrapper)!;
    classContentTest(editableWrapper, wrapperNoBorder, false);
  });
  test('should render when set placeholder', () => {
    const placeholderContent = 'input something';
    const { container } = render(<EditableArea placeholder={placeholderContent} />);
    const editableWrapper = container.querySelector(wrapper)!;
    textContentTest(editableWrapper.querySelector(content)!, placeholderContent);
  });
  test('should render when set width', () => {
    const { container } = render(<EditableArea width={100} />);
    styleTest(container.querySelector(wrapper)!, 'width: 100px;');
  });
  test('should render when content more than one line', () => {
    const { container } = render(<EditableArea width={100} />);
    fireEvent.click(container.querySelector(place)!);
    const editablePopup = document.querySelector(popup)!;
    classContentTest(editablePopup, popupShow);
    fireEvent.change(document.querySelectorAll('textarea')[0], { target: { value: 'value\n' } });
    fireEvent.blur(document.querySelectorAll('textarea')[0]);
    classContentTest(editablePopup, popupShow, false);
    textContentTest(container.querySelector(content)!, 'value...');
  });
  test('should render when set trim', () => {
    const { container } = render(<EditableArea trim />);
    changeValueHandle(container, ' value \n');
    textContentTest(container.querySelector(content)!, 'value');
  });
  // TODO
  test('should render when set getPopupContainer', () => {
    render(<EditableAreaContainer />);
    // TODO: should container.querySelectorAll, is a bug, but no problem on the browser
    expect(document.querySelectorAll('textarea').length).toBe(2);
  });
  test('should render when set maxHeight', () => {
    const maxHeight = 200;
    render(<EditableArea maxHeight={maxHeight} />);
    styleContentTest(document.querySelectorAll('textarea')[0], `max-height: ${maxHeight}px;`);
  });
  test('should render when set renderFooter', () => {
    const { container } = render(<EditableAreaRenderFooter />);
    fireEvent.click(container.querySelector(place)!);
    textContentTest(document.querySelector(textareaFooterClassName)!, 'Tip');
  });
});
describe('EditableArea[Value]', () => {
  test('should render when set defaultValue', () => {
    const { container } = render(<EditableArea defaultValue={defaultValue} />);
    testValue(container, defaultValue);
  });
  test('should render when set value', () => {
    const { container } = render(<EditableArea value={value} />);
    testValue(container, value);
    changeValueHandle(container, 'test\n');
    testValue(container, value);
  });
  test('should render when set value and defaultValue at the same time', () => {
    const { container } = render(<EditableArea defaultValue={defaultValue} value={value} />);
    testValue(container, value);
  });
  test('should render when set clearable', () => {
    const { container, rerender } = render(<EditableArea clearable />);
    expect(container.querySelector(clear)).not.toBeInTheDocument();
    changeValueHandle(container, 'value\n');
    expect(container.querySelector(clear)).toBeInTheDocument();
    rerender(<EditableArea clearable value='hello' />);
    expect(container.querySelector(clear)).toBeInTheDocument();
    fireEvent.click(container.querySelector(clear)!);
    textContentTest(container.querySelector(content)!, '');
    classContentTest(document.querySelector(popup)!, popupShow);
  });
  test('should render when set renderResult', () => {
    const { container } = render(<EditableArea renderResult={(v: string) => v + 1} />);
    changeValueHandle(container, value);
    textContentTest(container.querySelector(content)!, value + 1);
    textContentTest(document.querySelectorAll('textarea')[0], value);
    textContentTest(document.querySelectorAll('textarea')[1], value);
  });
  test('should render when is controlled', () => {
    const { container } = render(<EditableAreaControlled />);
    changeValueHandle(container, value);
    testValue(container, value);
  });
});
describe('EditableArea[Event]', () => {
  test('should render when set diabled', () => {
    const { container } = render(<EditableArea disabled />);
    const editableArea = container.querySelector(wrapper)!;
    classContentTest(editableArea, wrapperDisabled);
    fireEvent.click(container.querySelector(place)!);
    classContentTest(document.querySelector(popup)!, popupShow, false);
  });
  test('should render when set beforeChange', () => {
    const { container } = render(<EditableArea beforeChange={(v: string) => v + 1} />);
    changeValueHandle(container, 'value');
    testValue(container, 'value1');
  });
  test('should render when set onChange', () => {
    const changeFn = jest.fn();
    const { container } = render(<EditableArea onChange={changeFn} />);
    changeValueHandle(container, 'value');
    expect(changeFn.mock.calls.length).toBe(1);
  });
  test('should render when set onBlur', () => {
    const blurFn = jest.fn();
    const { container } = render(<EditableArea onBlur={blurFn} />);
    changeValueHandle(container, 'value');
    expect(blurFn.mock.calls.length).toBe(1);
  });
  test('should render when set onFocus', () => {
    const focusFn = jest.fn();
    const { container, rerender } = render(<EditableArea onFocus={focusFn} />);
    fireEvent.click(container.querySelector(place)!);
    fireEvent.click(document.querySelectorAll('textarea')[0]);
    expect(focusFn.mock.calls.length).toBe(1);
    rerender(<EditableArea onFocus={focusFn} />);
    fireEvent.focus(container.querySelector(place)!);
    classContentTest(document.querySelector(popup)!, popupShow);
  });
});
