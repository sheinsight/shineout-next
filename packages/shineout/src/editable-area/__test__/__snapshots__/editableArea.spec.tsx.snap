// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableArea[Base] should render correctly  1`] = `
<div
  class="so-editableArea-wrapper"
>
  <div
    class="so-editableArea-place so-editableArea-wrapper-padding-box so-editableArea-wrapper-inner-title-top so-editableArea-wrapper-inner-title-bottom"
    tabindex="0"
  >
    <div
      class="so-editableArea-content"
    >
      <span
        class="so-editableArea-placeholder"
      >
        input something
      </span>
    </div>
  </div>
</div>
`;

exports[`EditableArea[Base] should render correctly  2`] = `
<div
  class="so-editableArea-wrapper"
  style="width: 400px;"
>
  <div
    class="so-editableArea-place so-editableArea-wrapper-padding-box so-editableArea-wrapper-inner-title-top so-editableArea-wrapper-inner-title-bottom"
    tabindex="0"
  >
    <div
      class="so-editableArea-content"
    >
      <span
        class="so-editableArea-placeholder"
      >
        Input something
      </span>
    </div>
  </div>
</div>
`;

exports[`EditableArea[Base] should render correctly  3`] = `
<div
  id="popup-target"
  style="height: 200px; overflow: auto; position: relative; padding: 10px;"
>
  <div
    style="height: 100px;"
  />
  <div
    class="so-editableArea-wrapper"
  >
    <div
      class="so-editableArea-place so-editableArea-wrapper-padding-box so-editableArea-wrapper-inner-title-top so-editableArea-wrapper-inner-title-bottom"
      tabindex="0"
    >
      <div
        class="so-editableArea-content"
      >
        <span
          class="so-editableArea-placeholder"
        >
          scroll in container
        </span>
      </div>
    </div>
  </div>
  <div
    style="height: 140px;"
  />
</div>
`;

exports[`EditableArea[Base] should render correctly  4`] = `
<div>
  <div
    class="so-editableArea-wrapper"
  >
    <div
      class="so-editableArea-place so-editableArea-wrapper-padding-box so-editableArea-wrapper-inner-title-top so-editableArea-wrapper-inner-title-bottom"
      tabindex="0"
    >
      <div
        class="so-editableArea-content"
      >
        <span
          class="so-editableArea-placeholder"
        >
          input something
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`EditableArea[Base] should render correctly  5`] = `
<div>
  <div
    class="so-editableArea-wrapper"
  >
    <div
      class="so-editableArea-place so-editableArea-wrapper-padding-box so-editableArea-wrapper-inner-title-top so-editableArea-wrapper-inner-title-bottom"
      tabindex="0"
    >
      <div
        class="so-editableArea-content"
      >
        <div>
          Guessing what I inputed ?
        </div>
      </div>
    </div>
  </div>
</div>
`;
